src/pg_imports.rs

```diff
@@ -0,0 +1,110 @@
+//! PostrgeSQL and thus require conditional compilation.
+
+use pgx::*;
+
+#[derive(Default, Debug)]
+pub struct FuncDetail {
+    pub func_oid: pg_sys::Oid,
+    pub ret_type_oid: pg_sys::Oid,
+    pub retset: bool,
+    pub nvargs: ::std::os::raw::c_int,
+    pub vatype: pg_sys::Oid,
+    pub code: pg_sys::FuncDetailCode,
+}
+
+#[cfg(not(any(feature = "pg12", feature = "pg13")))]
+#[inline]
+pub fn func_get_detail<'a, I>(func_path: I, types: &mut [pg_sys::Oid]) -> FuncDetail
+where
+    I: IntoIterator<Item = &'a str>,
+{
+    let arg_cnt = types.len() as i32;
+    let fully_qualified_name = crate::build_pg_list_of_strings(func_path);
+    let mut true_typeoids: *mut pg_sys::Oid = std::ptr::null_mut();
+    let mut fd_struct: FuncDetail = Default::default();
+    fd_struct.code = unsafe {
+        pg_sys::func_get_detail(
+            fully_qualified_name.into_pg(),
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+            arg_cnt,
+            types.as_mut_ptr(),
+            false,
+            false,
+            false,
+            &mut fd_struct.func_oid,
+            &mut fd_struct.ret_type_oid,
+            &mut fd_struct.retset,
+            &mut fd_struct.nvargs,
+            &mut fd_struct.vatype,
+            &mut true_typeoids,
+            std::ptr::null_mut(),
+        )
+    };
+    fd_struct
+}
+
+// When compiling against PG12 and PG13 the underlying function takes fewer arguments
+#[cfg(any(feature = "pg12", feature = "pg13"))]
+#[inline]
+pub fn func_get_detail<'a, I>(func_path: I, types: &mut [pg_sys::Oid]) -> FuncDetail
+where
+    I: IntoIterator<Item = &'a str>,
+{
+    let arg_cnt = types.len() as i32;
+    let fully_qualified_name = crate::build_pg_list_of_strings(func_path);
+    let mut true_typeoids: *mut pg_sys::Oid = std::ptr::null_mut();
+    let mut fd_struct: FuncDetail = Default::default();
+    fd_struct.code = unsafe {
+        pg_sys::func_get_detail(
+            fully_qualified_name.into_pg(),
+            std::ptr::null_mut(),
+            std::ptr::null_mut(),
+            arg_cnt,
+            types.as_mut_ptr(),
+            false,
+            false,
+            &mut fd_struct.func_oid,
+            &mut fd_struct.ret_type_oid,
+            &mut fd_struct.retset,
+            &mut fd_struct.nvargs,
+            &mut fd_struct.vatype,
+            &mut true_typeoids,
+            std::ptr::null_mut(),
+        )
+    };
+    fd_struct
+}
+
+#[cfg(not(any(feature = "pg12", feature = "pg13")))]
+#[inline]
+pub fn set_sa_hashfuncid(
+    scalar_array_op: &mut pg_sys::ScalarArrayOpExpr,
+    hash_func_oid: pg_sys::Oid,
+) {
+    scalar_array_op.hashfuncid = hash_func_oid;
+}
+
+#[cfg(any(feature = "pg12", feature = "pg13"))]
+#[inline]
+pub fn set_sa_hashfuncid(
+    _scalar_array_op: &mut pg_sys::ScalarArrayOpExpr,
+    _hash_func_oid: pg_sys::Oid,
+) {
+    // the field didn't exist prior to pg14
+}
+
+// pg_guard doesn't compile, so we have to do without it for now.
```

Comment:
> Hmm how bad is this? Can it crash or does this behaviour become undefined?

I think it's benign for our use case. Our extension doesn't have any global state. IIRC when C-functions crash they bring down the entire PostgreSQL backend. So, I don't expect to run into undefined behaviour because of it.



> Do we know why it doesn't compile?

I've tried asking in the pgx discord but didn't get a definitive answer.

https://discord.com/channels/561648697805504526/749454413407584258/964516348711542824